### PR TITLE
Discern between VIRTUAL and ABSTRACT class bindings

### PR DESCRIPTION
**Pull request by reduz**
* Previous "virtual" classes (which can't be instantiated) are now correctly named "abstract" (GDREGISTER_ABSTRACT_CLASS).
* Added a new "virtual" category for classes, they can't be instantiated from the editor, but can be inherited from script and extensions (GDREGISTER_VIRTUAL_CLASS).
* Converted a large amount of classes from "abstract" to "virtual" where it makes sense, and added the relevant GDVIRTUAL bindings so they can be extended via extension.

Most classes (that make sense) have been converted. Missing:

* Physics servers
* VideoStream
* Script* classes.

which will go in separate PRs due to the complexity involved.

### CODE REVIEW

**Comment by touilleMan** [comment]
Wow this is super cool ! 😃 

> Script* classes.

Does it means the plan is to expose `ScriptServer/Script/ScriptInstance` to ClassDB and hence to allow them to be used from GDExtension ? (this is what point 1 of https://github.com/godotengine/godot-proposals/issues/3927 was about, so I'm super happy if this is happening ^^)

File: core/io/resource.cpp [hunk 1]
```diff
@@ -428,6 +443,11 @@ void Resource::_bind_methods() {
 	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "resource_local_to_scene"), "set_local_to_scene", "is_local_to_scene");
 	ADD_PROPERTY(PropertyInfo(Variant::STRING, "resource_path", PROPERTY_HINT_NONE, "", PROPERTY_USAGE_EDITOR), "set_path", "get_path");
 	ADD_PROPERTY(PropertyInfo(Variant::STRING_NAME, "resource_name"), "set_name", "get_name");
+
+	MethodInfo get_rid_bind("_get_rid");
+	get_rid_bind.return_val.type = Variant::RID;
+
+	::ClassDB::add_virtual_method(get_class_static(), get_rid_bind, true, Vector<String>(), true);
```
Comment (akien-mga) [inline core/io/resource.cpp:450]: Just curious, does it actually need the `::` scope here? Just above there's `ClassDB::bind_method` without extra scoping.

File: core/object/class_db.cpp [hunk 2]
```diff
@@ -557,6 +557,19 @@ bool ClassDB::can_instantiate(const StringName &p_class) {
...
+	OBJTYPE_RLOCK;
+
+	ClassInfo *ti = classes.getptr(p_class);
+	ERR_FAIL_COND_V_MSG(!ti, false, "Cannot get class '" + String(p_class) + "'.");
+#ifdef TOOLS_ENABLED
+	if (ti->api == API_EDITOR && !Engine::get_singleton()->is_editor_hint()) {
+		return false;
+	}
+#endif
+	return (!ti->disabled && ti->creation_func != nullptr && !(ti->native_extension && !ti->native_extension->create_instance) && ti->is_virtual);
```
Comment (akien-mga) [inline core/object/class_db.cpp:570]: I see you adapted this from `can_instantiate()` so this is basically `can_instantiate() && is_virtual`.

Just wondering if all checks from `can_instantiate()` are indeed wanted - e.g. do we want to exclude classes without creation func or disabled from being qualified virtual? Just a theoretical question.

File: core/register_core_types.cpp [hunk 3]
```diff
@@ -141,42 +141,42 @@ void register_core_types() {
...
 	GDREGISTER_CLASS(InputEventPanGesture);
 	GDREGISTER_CLASS(InputEventMIDI);
 
 	// Network
-	GDREGISTER_VIRTUAL_CLASS(IP);
+	GDREGISTER_ABSTRACT_CLASS(IP);
 
-	GDREGISTER_VIRTUAL_CLASS(StreamPeer);
+	GDREGISTER_ABSTRACT_CLASS(StreamPeer);
 	GDREGISTER_CLASS(StreamPeerExtension);
```
Comment (akien-mga) [inline core/register_core_types.cpp:174]: `StreamPeer`, `PacketPeer` and `MultiplayerPeer` all have an `*Extension` wrapper which AFAIU was specifically added to workaround the limitation that this PR lifts. So these could likely be registered as virtual instead, and the `*Extension` wrappers dropped.

But that's probably stuff for a future PR so it might be good to keep it as is in this one. CC @Faless

File: core/register_core_types.cpp [hunk 4]
```diff
@@ -226,19 +226,19 @@ void register_core_types() {
...
 	GDREGISTER_CLASS(PackedDataContainer);
-	GDREGISTER_VIRTUAL_CLASS(PackedDataContainerRef);
+	GDREGISTER_ABSTRACT_CLASS(PackedDataContainerRef);
 	GDREGISTER_CLASS(AStar);
 	GDREGISTER_CLASS(AStar2D);
 	GDREGISTER_CLASS(EncodedObjectAsID);
 	GDREGISTER_CLASS(RandomNumberGenerator);
 
-	GDREGISTER_VIRTUAL_CLASS(ResourceImporter);
+	GDREGISTER_ABSTRACT_CLASS(ResourceImporter);
```
Comment (akien-mga) [inline core/register_core_types.cpp:235]: Isn't that something that might typically make sense to implement as virtual?

**Comment by reduz** [comment]
@touilleMan I think ScriptInstance will be exposed via a native extension API, since its more efficient it does not use the binder,  but Script and ScriptLanguage will be exposed as regular ClassDB classes. Likely as ScriptExtension and ScriptLanguageExtension.

File: doc/classes/Texture2D.xml [hunk 5]
```diff
@@ -12,6 +12,58 @@
 	<tutorials>
 	</tutorials>
 	<methods>
+		<method name="_draw" qualifiers="virtual const">
+			<return type="void" />
```
Comment (akien-mga) [inline doc/classes/Texture2D.xml:16]: Just a note in passing but as we recently moved the docs for constructors and operators to dedicated `<constructors>` and `<operators>` sections, it might make sense to do the same (in a follow-up) with a `<virtuals>` or something like this.

So we can make the docs a bit clearer with a separation between the API users can use, and the one users can extend/reimplement.

File: editor/editor_node.cpp [hunk 6]
```diff
@@ -3858,18 +3858,18 @@ void EditorNode::register_editor_types() {
 	GDREGISTER_CLASS(EditorScript);
 	GDREGISTER_CLASS(EditorSelection);
 	GDREGISTER_CLASS(EditorFileDialog);
-	GDREGISTER_VIRTUAL_CLASS(EditorSettings);
+	GDREGISTER_ABSTRACT_CLASS(EditorSettings);
```
Comment (akien-mga) [inline editor/editor_node.cpp:3861]: I'm not sure I understand why these editor types are defined as abstract (and were defined as virtual before). What's abstract in `EditorSettings` or `EditorResourcePreview`? Is it just to prevent them from being instantiated?

File: scene/gui/spin_box.cpp [hunk 7]
```diff
@@ -39,7 +39,7 @@ Size2 SpinBox::get_minimum_size() const {
 	return ms;
 }
 
-void SpinBox::_value_changed(double) {
+void SpinBox::_value_changed(double p_value) {
 	String value = TS->format_number(String::num(get_value(), Math::range_step_decimals(get_step())));
```
Comment (akien-mga) [inline scene/gui/spin_box.cpp:43]: Could be material for another PR, but seems weird to rely on `get_value()` here instead of actually using the value passed to the method.

File: scene/register_scene_types.cpp [hunk 8]
```diff
@@ -765,7 +765,7 @@ void register_scene_types() {
 	GDREGISTER_CLASS(RibbonTrailMesh);
 	GDREGISTER_CLASS(PointMesh);
 	GDREGISTER_VIRTUAL_CLASS(Material);
-	GDREGISTER_VIRTUAL_CLASS(BaseMaterial3D);
+	GDREGISTER_ABSTRACT_CLASS(BaseMaterial3D);
```
Comment (akien-mga) [inline scene/register_scene_types.cpp:768]: Could maybe make sense as virtual? Either way, all this can be discussed in follow-ups if/when the needs arise and then changing it is just a macro rename away.

File: scene/resources/material.cpp [hunk 9]
```diff
@@ -88,6 +88,37 @@ void Material::inspect_native_shader_code() {
 	}
 }
 
+RID Material::get_shader_rid() const {
+	RID ret;
+	if (GDVIRTUAL_CALL(_get_shader_rid, ret)) {
+		return ret;
+	}
+	ERR_FAIL_V_MSG(RID(), "Unimplemented function");
```
Comment (akien-mga) [inline scene/resources/material.cpp:96]: Same for all other messages (adding "virtual" and dot at end of sentence):
```suggestion
	ERR_FAIL_V_MSG(RID(), "Unimplemented virtual function.");
```

File: scene/resources/mesh.cpp [hunk 10]
```diff
@@ -40,6 +40,125 @@
...
+	ERR_FAIL_V_MSG(0, "Unimplemented function");
+}
+
+StringName Mesh::get_blend_shape_name(int p_index) const {
+	StringName ret;
+	if (GDVIRTUAL_CALL(_get_blend_shape_name, p_index, ret)) {
+		return ret;
+	}
+
+	ERR_FAIL_V_MSG("Unimplemented function", "");
```
Comment (akien-mga) [inline scene/resources/mesh.cpp:142]: You swapped the error cond and the message :)

File: scene/resources/texture.cpp [hunk 11]
```diff
@@ -38,23 +38,61 @@
 #include "scene/resources/bit_map.h"
 #include "servers/camera/camera_feed.h"
 
+int Texture2D::get_width() const {
+	int ret;
+	if (GDVIRTUAL_CALL(_get_width, ret)) {
+		return ret;
+	}
+	ERR_FAIL_V_MSG(0, "Unimplemented method");
```
Comment (akien-mga) [inline scene/resources/texture.cpp:46]: You used "function" in all other classes, would be good to do in Texture2D too for consistency.

File: scene/resources/texture.h [hunk 12]
```diff
@@ -474,15 +492,21 @@ class Texture3D : public Texture {
...
-	TypedArray<Image> _get_data() const;
-
-public:
-	virtual Image::Format get_format() const = 0;
-	virtual int get_width() const = 0;
-	virtual int get_height() const = 0;
-	virtual int get_depth() const = 0;
-	virtual bool has_mipmaps() const = 0;
-	virtual Vector<Ref<Image>> get_data() const = 0;
+	TypedArray<Image> _get_datai() const;
```
Comment (akien-mga) [inline scene/resources/texture.h:495]: What does `i` stand for here?

File: servers/audio/audio_stream.cpp [hunk 13]
```diff
@@ -105,6 +105,30 @@ void AudioStreamPlaybackResampled::_begin_resample() {
...
+	}
+	WARN_PRINT_ONCE("AudioStreamResampled::_mix_resampled unimplemented!");
+	return 0;
+}
+float AudioStreamPlaybackResampled::get_stream_sampling_rate() {
+	float ret;
+	if (GDVIRTUAL_CALL(_get_stream_sampling_rate, ret)) {
+		return ret;
+	}
+	WARN_PRINT_ONCE("AudioStreamResampled::_mix_resampled unimplemented!");
```
Comment (akien-mga) [inline servers/audio/audio_stream.cpp:121]:
```suggestion
	WARN_PRINT_ONCE("AudioStreamResampled::_get_stream_sampling_rate unimplemented!");
```

Though maybe this was the first one you changed before settling on `ERR_FAIL_V_MSG(0, "Unimplemented function");`?

File: servers/register_server_types.cpp [hunk 14]
```diff
@@ -110,7 +110,7 @@ void preregister_server_types() {
 	shader_types = memnew(ShaderTypes);
 
 	GDREGISTER_CLASS(TextServerManager);
-	GDREGISTER_VIRTUAL_CLASS(TextServer);
+	GDREGISTER_ABSTRACT_CLASS(TextServer);
 	GDREGISTER_CLASS(TextServerExtension);
```
Comment (akien-mga) [inline servers/register_server_types.cpp:114]: Another candidate for merging `TextServerExtension` into a virtual `TextServer` I guess @bruvzg.

**Review by akien-mga — COMMENTED** [review]
Looks pretty good to me, some comments and nitpicks.

**Comment by akien-mga** [comment]
Looking at the kind of `GDVIRTUAL_CALL` + `ERR_FAIL_MSG` construct which happens a lot for virtual methods which **must** be implemented, I was thinking that it could make sense to add a new `GDVIRTUAL_CALL_REQUIRED` that would handle this automatically with an explicit error message.

So this:
```cpp
Ref<Material> Mesh::surface_get_material(int p_idx) const {
	Ref<Material> ret;
	if (GDVIRTUAL_CALL(_surface_get_material, p_idx, ret)) {
		return ret;
	}

	ERR_FAIL_V_MSG(Ref<Material>(), "Unimplemented function");
}
```

Would become this:
```cpp
Ref<Material> Mesh::surface_get_material(int p_idx) const {
	Ref<Material> ret;
	if (GDVIRTUAL_CALL_REQUIRED(_surface_get_material, p_idx, ret)) {
		return ret;
	}
	return Ref<Material>();
}
```

And `GDVIRTUAL_CALL_REQUIRED` would use `##m_name##` and `get_class()` to compose a nice error message when it's not implemented, referring to the name of the virtual methods that must be implemented.
So you'd get an error like this:
```
Missing implementation for required virtual method '_surface_get_material' in Mesh.
```

**Comment by akien-mga** [comment]
For the record, `godot-cpp` fails building after this PR with:
```
g++ -o src/register_types.os -c -std=c++17 -fPIC -Wwrite-strings -Og -g -m64 -fPIC -DDEBUG_ENABLED -DDEBUG_METHODS_ENABLED -I. -Igodot-headers -Iinclude -Igen/include -I/home/runner/work/godot/godot/godot-cpp/gen/include -I/home/runner/work/godot/godot/godot-cpp/include -I/home/runner/work/godot/godot/godot-cpp/godot-headers -Isrc src/register_types.cpp
In file included from src/register_types.cpp:35:
/home/runner/work/godot/godot/godot-cpp/include/godot_cpp/core/class_db.hpp: In instantiation of ‘static void godot::ClassDB::register_class() [with T = ExampleRef]’:
src/register_types.cpp:44:38:   required from here
/home/runner/work/godot/godot/godot-cpp/include/godot_cpp/core/class_db.hpp:162:14: error: invalid conversion from ‘void*’ to ‘GDNativeExtensionClassGetRID’ {aka ‘long unsigned int (*)(void*)’} [-fpermissive]
  162 |   (void *)cl.name, // void *class_userdata;
      |           ~~~^~~~
      |              |
      |              void*
/home/runner/work/godot/godot/godot-cpp/include/godot_cpp/core/class_db.hpp: In instantiation of ‘static void godot::ClassDB::register_class() [with T = Example]’:
src/register_types.cpp:45:35:   required from here
/home/runner/work/godot/godot/godot-cpp/include/godot_cpp/core/class_db.hpp:162:14: error: invalid conversion from ‘void*’ to ‘GDNativeExtensionClassGetRID’ {aka ‘long unsigned int (*)(void*)’} [-fpermissive]
```

We need to disable the godot-cpp test in this PR temporarily until this is merged and a matching counterpart is merged in godot-cpp.

**Comment by akien-mga** [comment]
For the record, here's the list of classes currently exposed as virtual as of this PR (it's not visible in the diff due to the name swap with virtual -> abstract so those didn't get a diff):
```
$ rg GDREGISTER_VIRTUAL_CLASS
core/object/class_db.h
442:#define GDREGISTER_VIRTUAL_CLASS(m_class)            \

servers/register_server_types.cpp
142:    GDREGISTER_VIRTUAL_CLASS(AudioStreamPlaybackResampled);
145:    GDREGISTER_VIRTUAL_CLASS(AudioEffect);
146:    GDREGISTER_VIRTUAL_CLASS(AudioEffectInstance);

scene/register_scene_types.cpp
320:    GDREGISTER_VIRTUAL_CLASS(BaseButton);
341:    GDREGISTER_VIRTUAL_CLASS(Range);
462:    GDREGISTER_VIRTUAL_CLASS(VisualInstance3D);
463:    GDREGISTER_VIRTUAL_CLASS(GeometryInstance3D);
748:    GDREGISTER_VIRTUAL_CLASS(Mesh);
756:    GDREGISTER_VIRTUAL_CLASS(PrimitiveMesh);
767:    GDREGISTER_VIRTUAL_CLASS(Material);
808:    GDREGISTER_VIRTUAL_CLASS(Texture);
809:    GDREGISTER_VIRTUAL_CLASS(Texture2D);
822:    GDREGISTER_VIRTUAL_CLASS(TextureLayered);
824:    GDREGISTER_VIRTUAL_CLASS(Texture3D);
845:    GDREGISTER_VIRTUAL_CLASS(StyleBox);
```

File: servers/register_server_types.cpp (hunk 14 above)
Comment (bruvzg) [reply servers/register_server_types.cpp:114]: This make sense, but last time I have tried to do so, every `GDVIRTUAL` binding which is using enums for the `TextServer` stopped working, seems like registering enums and binding virtual methods should have a certain order. So it might need some extra changes.

File: core/register_core_types.cpp (hunk 4 above)
Comment (reduz) [reply core/register_core_types.cpp:235]: Yes, I completely forgot about these, I will add them.

Comment (reduz) [reply core/register_core_types.cpp:235]: Looking at them, this will require likely another PR because its a lot of code.

File: servers/register_server_types.cpp (hunk 14 above)
Comment (akien-mga) [reply servers/register_server_types.cpp:114]: It also sounds like @reduz suggests keeping `*Extension` classes for classes where *a lot* of the API is virtual, so this might be the case here.

File: scene/resources/texture.h (hunk 12 above)
Comment (reduz) [reply scene/resources/texture.h:495]: internal I suppose, its not user exposed and _get_data is virtual already.

File: servers/register_server_types.cpp (hunk 14 above)
Comment (reduz) [reply servers/register_server_types.cpp:114]: I think for the large ones (like script, physics server, etc. Using Extension makes more sense because otherwise the code gets to be a mess.

File: .github/workflows/linux_builds.yml [hunk 15]
```diff
@@ -36,7 +36,9 @@ jobs:
             tests: true
             sconsflags: float=64 use_asan=yes use_ubsan=yes
             proj-test: true
-            godot-cpp-test: true
+            # Can be turned off for PRs that intentionally break compat with godot-cpp,
+            # until both the upstream PR and the matching godot-cpp changes are merged.
+            godot-cpp-test: false
```
Comment (akien-mga) [inline .github/workflows/linux_builds.yml:41]: Trailing whitespace.

**Review by akien-mga — APPROVED** [review]

File: core/object/class_db.cpp (hunk 2 above)
Comment (reduz) [reply core/object/class_db.cpp:570]: Good point, I am not sure. For the time being I only made this changes to affect the editor, but not more. We can improve it towards the future I suppose.

File: core/register_core_types.cpp (hunk 3 above)
Comment (reduz) [reply core/register_core_types.cpp:174]: Extension makes sense if its something very large, so I guess it is up to Fabio to change.

**Comment by akien-mga** [comment]
Thanks!
